docs/proposals/volumes.md

```diff
@@ -0,0 +1,373 @@
+
+Relabeling should be an optional aspect of a volume plugin to accomodate:
+
+1.  volume types for which generalized relabeling support is not sufficient
+2.  testing for each volume plugin individually
+
+#### SELinux and distributed file systems
+
+Some distributed filesystems have complications where SELinux is involved.  Let's look at NFS as an
+example.  Files in NFS volumes cannot have their contexts changed with `chcon` after the volume is
+mounted; the `context` argument must be passed to the `mount` call.
+
+For NFS, there is an SELinux boolean, `virt_use_nfs` that allows containers to use files with the
+SELinux type `nfs_t`.  Setting `virt_use_nfs` on a system would allow containers to use *any* NFS
+mount without the kubelet having to relabel the files.  However, the semantics would be different
+from what might be expected, since SELinux will not provide any protection against containers from
+one pod using files in an NFS mount belonging to another pod.
+
+NFS is a case where the `chcon` approach isn't going to solve the whole problem.  Instead, the NFS
+volume plugin will have to handle calling `mount` with the `context` argument in order to gain
+cross-container protection from SELinux.
+
+Not all distributed file systems support SELinux uniformly.  Some file systems provide only the
+coarse-level control via an SELinux boolean to determine whether containers can use them.  The
+internal API should account for this.
+
+## Proposed Design
+
+Our proposed design should minimize code for handling ownership and SELinux required in:
+
+1.  volume plugins
+2.  the kubelet
+
+### API changes
+
+There are a couple factors to consider for the API:
+
+1.  Currently, it is possible to specify a different SELinux context for each container in a pod
+2.  There must be a way to specify a supplemental group that all containers in a pod are part of
+3.  There should be a uniform way to specify in the API that a volume requires ownership / label
+    management
+
+As stated in the assumptions of this proposal, we will not secure containers in a pod from one
+another.  This greatly simplifies the story for SELinux -- we can make the SELinux context uniform
+across the containers in a pod and remove it from the container level entirely.  The pod-level
+security proposal deals with this in detail.
+
+The pod-level supplemental group is necessary to capture groups that cut across all containers in a
+pod, which is the simplest way to guarantee that all arbitrary combinations of UID/GIDs in
+containers can share a volume.  The pod-level security context proposal also deals with this change
+in detail.
+
+Finally, we must have a clear way for users to indicate that Kubernetes should manage ownership and
+labeling of a volume, when it is possible.  For this, the `VolumeSource` should have a new field,
+`Manage`, and the volume `Builder` interface have two new methods, `RequiresOwnershipManagement()`
+and `RequiresLabelManagement()`.
```

Comment:
I would prefer to gate on container-wide capability after some additional thought.  I don't want a validation bug to mean that pods get incomplete SELinux protection at runtime.  Remember that docker may allocate a unique SELinux context for each container anyway if a pod slips by with a blank SELinux context that it shouldn't have and the Kubelet doesn't perform the duties it should for SELinux.  I would prefer to fail fast instead of starting a pod in the wrong configuration.